components/webui/client/src/pages/IngestPage/Compress/PathsSelectFormItem.tsx

```diff
@@ -0,0 +1,141 @@
+import React, {
+    useCallback,
+    useRef,
+    useState,
+} from "react";
+
+import {
+    MinusOutlined,
+    PlusOutlined,
+} from "@ant-design/icons";
+import {FileEntry} from "@webui/common/schemas/os";
+import {
+    Form,
+    message,
+    TreeSelect,
+} from "antd";
+
+import {listFiles} from "../../../api/os";
+import {
+    addServerPrefix,
+    ROOT_PATH,
+    toTreeNode,
+    type TreeNode,
+} from "./utils";
+
+
+/**
+ * Height of the dropdown list in pixels.
+ */
+const LIST_HEIGHT_PX = 512;
+
+
+/**
+ * Root tree node representing the filesystem root.
+ */
+const ROOT_NODE: TreeNode = Object.freeze({
+    id: ROOT_PATH,
+    isLeaf: false,
+    pId: null,
+    title: ROOT_PATH,
+    value: ROOT_PATH,
+});
+
+/**
+ * Icon component for tree node expand/collapse state.
+ *
+ * @param props Component props.
+ * @param props.expanded Whether the node is expanded (passed by Ant Design).
+ * @return Plus or minus icon based on expand state.
+ */
+// The "expanded" prop name is dictated by Ant Design TreeSelect.
+// eslint-disable-next-line react/boolean-prop-naming
+const SwitcherIcon = ({expanded}: {expanded?: boolean}) => (expanded ?
+    <MinusOutlined style={{color: "grey"}}/> :
+    <PlusOutlined style={{color: "grey"}}/>);
+
+/**
+ * Form item with TreeSelect for selecting file paths for compression.
+ * Supports lazy loading and search with auto-expand.
+ *
+ * @return Form.Item component with TreeSelect.
+ */
+const PathsSelectFormItem = () => {
+    const [treeData, setTreeData] = useState<TreeNode[]>([ROOT_NODE]);
+    const [expandedKeys, setExpandedKeys] = useState<string[]>([]);
+    const loadedPaths = useRef(new Set<string>());
+
+    const addNodes = useCallback((nodes: TreeNode[]) => {
+        setTreeData((prev) => {
+            const existingIds = new Set(prev.map((n) => n["id"] as string));
+            const newNodes = nodes.filter((n) => false === existingIds.has(n["id"] as string));
+
+            return 0 < newNodes.length ?
+                [
+                    ...prev,
+                    ...newNodes,
+                ] :
+                prev;
+        });
+    }, []);
+
+    const loadPath = useCallback(async (path: string) => {
+        if (loadedPaths.current.has(path)) {
+            return;
+        }
+        loadedPaths.current.add(path);
```

Comment:
note we need to track the item BEFORE the request, or this guard cannot prevent multiple `/os/ls` requests getting fired at the same time